consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,241 @@
+	r := interactor
+	if interactor == nil {
+		r = &defaultRootchainInteractor{}
+	}
+
+	return &checkpointManager{
+		sender:            sender,
+		blockchain:        blockchain,
+		consensusBackend:  backend,
+		rootchain:         r,
+		checkpointsOffset: checkpointOffset,
+	}
+}
+
+// getCurrentCheckpointID queries CheckpointManager smart contract and retrieves current checkpoint id
+// (the latest checkpoint block number)
+func (c checkpointManager) getCurrentCheckpointID() (uint64, error) {
+	checkpointIDMethodEncoded, err := currentCheckpointIDMethod.Encode([]interface{}{})
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters: %w", err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain: %w", err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number: %w",
+			currentCheckpointID, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, isEndOfEpoch bool) error {
+	lastCheckpointBlockNumber, err := c.getCurrentCheckpointID()
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	initialBlockNumber := lastCheckpointBlockNumber + 1
+
+	var parentExtra *Extra
+
+	var parentHeader *types.Header
+
+	if initialBlockNumber < latestHeader.Number {
+		found := false
+		parentHeader, found = c.blockchain.GetHeaderByNumber(lastCheckpointBlockNumber)
+
+		if !found {
+			return fmt.Errorf("block %d was not found", lastCheckpointBlockNumber)
+		}
+
+		parentExtra, err = GetIbftExtra(parentHeader.ExtraData)
+		if err != nil {
+			return err
+		}
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNumber := initialBlockNumber; blockNumber < latestHeader.Number; blockNumber++ {
+		currentHeader, found := c.blockchain.GetHeaderByNumber(blockNumber)
+		if !found {
+			return fmt.Errorf("block %d was not found", blockNumber)
+		}
+
+		currentExtra, err := GetIbftExtra(currentHeader.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		parentEpochNumber := parentExtra.Checkpoint.EpochNumber
+		currentEpochNumber := currentExtra.Checkpoint.EpochNumber
+		parentHeader = currentHeader
+		parentExtra = currentExtra
+
+		// send pending checkpoints only for epoch ending blocks
+		if parentEpochNumber == currentEpochNumber {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *parentHeader, *parentExtra, true)
```

Comment:
Can this incur in very elevated fees if there are a lot of them?

Reply:
Yes you are right and I've mentioned it once in the RFC-30 comment. It would be better if we would be able to submit batch of checkpoints through a single transaction. However current proposal for checkpointing doesn't  have such concept of submitting multiple checkpoints at once, but one by one exclusively.

(cc @b00ris)

Reply:
We should throttle this I think.

Reply:
Interesting point. 

However, I would say we need separate brainstorming session(s) and RFC for any further optimizations. At first, we need to see is it likely for this to happen and then to agree on exact strategy how to mitigate it.

Reply:
Creating an issue to track this https://polygon.atlassian.net/browse/EVM-186?atlOrigin=eyJpIjoiM2ZmNTVjZmRhMDRkNDkzNzkxMDYwYjVjN2VkMjE4ODQiLCJwIjoiaiJ9